### PR TITLE
Weather implementation

### PR DESCRIPTION
**Pull request by qualterz**
## Description

An implementation of basic weather systems.

The weather component attached to a world instance would be handled for all clients, except those that have their own weather component; these clients would be handled separately.

## Test Plan

<details>

<summary>Playground</summary>

```rust
fn handle_command_events(
    instances: Query<Entity, With<Instance>>,
    mut exec_cmds: EventReader<CommandExecution>,
    mut commands: Commands,
) {
    for cmd in exec_cmds.iter() {
        let msg = cmd.command.to_string();
        let ent = instances.single();

        match msg.as_str() {
            "ar" => {
                commands.entity(ent).insert(Rain(WEATHER_LEVEL.end));
            }
            "rr" => {
                commands.entity(ent).remove::<Rain>();
            }
            "at" => {
                commands.entity(ent).insert(Thunder(WEATHER_LEVEL.end));
            }
            "rt" => {
                commands.entity(ent).remove::<Thunder>();
            }
            _ => (),
        };
    }
}
```

</details>

Steps:
1. Run `cargo test --package valence --lib -- weather::test`

#### Related
Part of #210
Past approach #106

### CODE REVIEW

File: crates/valence/src/instance.rs [hunk 1]
```diff
@@ -431,6 +434,46 @@ impl Instance {
...
+            kind: GameEventKind::EndRaining,
+            value: f32::default(),
+        });
+    }
+
+    /// Sets a level of the rain in the instance.
+    pub fn set_raining_level(&mut self, level: f32) {
+        self.write_packet(&GameStateChangeS2c {
+            kind: GameEventKind::RainLevelChange,
+            value: num::clamp(
```
Comment (qualterz) [inline crates/valence/src/instance.rs:458]: I have doubts about should the value be clamped or not.

File: crates/valence/src/weather.rs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    /// Should be between [`MIN_WEATHER_LEVEL`] and [`MAX_WEATHER_LEVEL`].
+    ///
+    /// The [`None`] value means no raining event.
+    pub raining: Option<f32>,
+    /// Contains the thunder level.
+    /// Should be between [`MIN_WEATHER_LEVEL`] and [`MAX_WEATHER_LEVEL`].
+    ///
+    /// The [`None`] value means no thunder.
+    pub thunder: Option<f32>,
+}
```
Comment (rj00a) [inline crates/valence/src/weather.rs:19]: Can you make this module a child of the `instance` module?

File: crates/valence/src/instance.rs (hunk 1 above)
Comment (qualterz) [reply crates/valence/src/instance.rs:458]: Also there is a core `clamp` function implementation, which should be used?

File: crates/valence/src/instance/weather.rs [hunk 3]
```diff
@@ -0,0 +1,337 @@
+use bevy_ecs::prelude::*;
+use bevy_ecs::schedule::SystemConfigs;
+use valence_protocol::packet::s2c::play::game_state_change::GameEventKind;
+use valence_protocol::packet::s2c::play::GameStateChangeS2c;
+
+use super::Instance;
+use crate::prelude::*;
+
+pub const WEATHER_LEVEL_MIN: f32 = 0_f32;
+pub const WEATHER_LEVEL_MAX: f32 = 1_f32;
```
Comment (dyc3) [inline crates/valence/src/instance/weather.rs:10]: Can this be a `Range` instead?

**Review by dyc3 — COMMENTED** [review]
Seems good. Fix the merge conflict and we can merge it

File: crates/valence/src/instance.rs [hunk 4]
```diff
@@ -28,6 +29,7 @@ use crate::view::ChunkPos;
 mod chunk;
 mod chunk_entry;
 mod paletted_container;
+pub mod weather;
```
Comment (rj00a) [inline crates/valence/src/instance.rs:32]: Keeping the weather module in the `instance` module doesn't really fit since it works on clients too. So it's original place is probably fine.

File: crates/valence/src/weather.rs [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+
+use bevy_ecs::prelude::*;
+use bevy_ecs::schedule::SystemConfigs;
+use valence_protocol::packet::s2c::play::game_state_change::GameEventKind;
+use valence_protocol::packet::s2c::play::GameStateChangeS2c;
+
+use crate::packet::WritePacket;
+use crate::prelude::*;
+
+pub const WEATHER_LEVEL_RANGE: Range<f32> = 0_f32..1_f32;
```
Comment (qualterz) [inline crates/valence/src/weather.rs:11]: Is not the name of this constant too wordy? Maybe just `WEATHER_LEVEL` would be better?

File: crates/valence/src/weather.rs [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+        instance.set_weather(weather);
+    });
+}
+
+fn handle_weather_for_joined_player(
+    mut clients: Query<&mut Client, Added<Client>>,
+    weathers: Query<&Weather, With<Instance>>,
+) {
+    clients.for_each_mut(|mut client| {
+        if let Ok(weather) = weathers.get_single() {
```
Comment (qualterz) [inline crates/valence/src/weather.rs:150]: How to get the weather component of the instance where the client is joined?

Comment (qualterz) [reply crates/valence/src/weather.rs]: Alright, I got the instance entity from `McEntity` using `Query<(&mut Client, &McEntity), Added<Client>>`, and I used the instance entity to get the weather component from `Query<&Weather, With<Instance>>`.

Comment (rj00a) [reply crates/valence/src/weather.rs]: You'll want to use the `Location` component instead.

File: crates/valence/src/weather.rs [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+            client.end_raining();
+        }
+    })
+}
+
+fn handle_weather_change_per_client(mut query: Query<(&mut Client, &Weather), Changed<Weather>>) {
+    query.for_each_mut(|(mut client, weather)| {
+        client.set_weather(weather);
+    });
+}
```
Comment (rj00a) [inline crates/valence/src/weather.rs:177]: What happens if an entity has both client components and an `Instance` component? Changes to `Weather` will be sent twice -- once for the client and once for the instance. Looks like you'll want to add `Without<Instance>` to these systems.

File: crates/valence/src/weather.rs [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+        handle_weather_for_joined_player,
+        handle_weather_begin_per_instance.before(handle_weather_change_per_instance),
+        handle_weather_change_per_instance.before(handle_weather_end_per_instance),
+        handle_weather_end_per_instance,
+        handle_weather_begin_per_client.before(handle_weather_change_per_client),
+        handle_weather_change_per_client.before(handle_weather_end_per_client),
+        handle_weather_end_per_client,
+    )
+        .into_configs()
+}
```
Comment (rj00a) [inline crates/valence/src/weather.rs:190]: Following #289, these systems should be wrapped up in a `Plugin`.

**Comment by qualterz** [comment]
My final step is to document weather module.

**Comment by qualterz** [comment]
Something weird with tests execution, in some cases they may fail. I'll just increase the number of ticks and cross my fingers that this will be a permanent fix.

**Comment by qualterz** [comment]
I don't know should the framework handle the `Thunder` component removal or not. I think would be better to leave this to a user and the weather components would be just an ECS API for weather events emitting.

**Comment by qualterz** [comment]
During my play tests, I preferred to leave the `Thunder` component handling for ergonomics.

**Comment by qualterz** [comment]
Not sure about scheduling but tests should pass. If there's no problems, the pull request can be merged.

**Review by dyc3 — APPROVED** [review]
